drivers/nhs-nut.c

```diff
@@ -0,0 +1,2070 @@
+                            } // end if
+                            else
+                                upsdebugx(1,"Checksum of pkt_hwinfo is corrupted or not initialized. Waiting for new request...\r\n");
+                        } // end else
+                    } // end if
+                } // end if
+            } // end if
+        } // end if
+        // Now the nobreak read buffer is empty. We need a hw info packet to discover several variables, like number of batteries, to calculate some data
+        if (!lastpkthwinfo.checksum_ok) {
+            upsdebugx(1,"pkt_hwinfo loss -- Requesting\r\n");
+            // if size == 0, packet maybe not initizated, then send a initialization packet to obtain data
+            // Send two times the extended initialization string, but, on fail, try randomly send extended or normal
+            if (send_extended < 6) {
+                upsdebugx(1,"Sending extended initialization packet. Try %d",send_extended+1);
+                bwritten = write_serial_int(serial_fd,string_initialization_long,9);
+                send_extended++;
+            } // end if
+            else {
+                // randomly send
+                if (rand() % 2 == 0) {
+                    upsdebugx(1,"Sending long initialization packet");
+                    bwritten = write_serial_int(serial_fd,string_initialization_long,9);
+                } // end if
+                else {
+                    upsdebugx(1,"Sending short initialization packet");
+                    bwritten = write_serial_int(serial_fd,string_initialization_short,9);
+                } // end else
+            } // end else
+            if (bwritten < 0) {
+                upsdebugx(1,"Problem to write data to %s",porta);
+                if (bwritten == -1) {
+                    upsdebugx(1,"\r\nData problem\r\n");
+                }
+                if (bwritten == -2) {
+                    upsdebugx(1,"\r\nFlush problem\r\n");
+                }
+                close(serial_fd);
+                serial_fd = -1;
+            } // end if
+            else {
+                if (checktime > max_checktime)
+                    checktime = max_checktime;
+                else {
+                    upsdebugx(1,"Increase checktime to %d",checktime + 10000);
+                    checktime = checktime + 10000;
+                }
+                usleep(checktime);
+            } // end else
+        } // end if
+    } // end else
+    upsdebugx(1,"End updateinfo()");
+}
+
+void upsdrv_shutdown(void) {
+    upsdebugx(1,"Start shutdown()");
+    if (serial_fd != -1) {
+        close(serial_fd);
+        serial_fd = -1;
+    }
+    upsdebugx(1,"Driver shutdown");
+}
+
+void upsdrv_cleanup(void) {
+    upsdebugx(1,"Start cleanup()");
+    if (serial_fd != -1) {
+        close(serial_fd);
+        serial_fd = -1;
+    }
+    upsdebugx(1,"End cleanup()");
+}
+
+void upsdrv_initups(void) {
+    upsdebugx(1,"Start initups()");
```

Comment:
`upsdrv_initups()` should take care of hardware communications init (maybe fail with `fatalx()` if `port` is not accessible).

The `upsdrv_initinfo()` method is called later to detect the device details (maybe fail with `fatalx()` if no compatible device was found) and set write-once values like device model etc. and populate the list of supported commands and/or settable variables, if any.

Please do check `docs/new-drivers.txt` for general driver code structure overview.